core/drivers/riscv_zkr_rng.c

```diff
@@ -4,46 +4,64 @@
  */
 
 #include <crypto/crypto.h>
+#include <encoding.h>
+#include <kernel/delay.h>
 #include <kernel/panic.h>
 #include <riscv.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 
-#define OPST_BIST 0b00
-#define OPST_WAIT 0b01
-#define OPST_ES16 0b10
-#define OPST_DEAD 0b11
+#define RNG_TIMEOUT_US	1000000
 
-TEE_Result hw_get_random_bytes(void *buf, size_t len)
+static inline bool __must_check seed_get_random_u16(uint16_t *val)
 {
-	uint8_t *ptr = buf;
-	uint32_t val = 0;
+	uint64_t timeout = timeout_init_us(RNG_TIMEOUT_US);
+	uint32_t seed = 0;
+	uint32_t opst = 0;
 
-	while (len > 0) {
+	do {
 		/*
-		 * The seed register must be accessed using CSR read-write
-		 * instructions. The write operation is ignored and serves
-		 * to indicate polling and flushing.
+		 * The seed register must be accessed using CSR
+		 * read-write instructions. The write operation
+		 * is ignored and serves to indicate polling and
+		 * flushing.
 		 */
-		val = swap_csr(CSR_SEED, val);
+		seed = swap_csr(CSR_SEED, 0);
+		opst = seed & SEED_OPST;
 
-		switch (val >> 30) {
-		case OPST_BIST:
-		case OPST_WAIT:
-			continue;
-		case OPST_ES16:
-			*ptr++ = val & 0xff;
-			len--;
-			if (len > 0) {
-				*ptr++ = val >> 8;
-				len--;
-			}
-			break;
-		case OPST_DEAD:
+		switch (opst) {
+		case SEED_OPST_ES16:
+			*val = seed & SEED_ENTROPY;
+			return true;
+		case SEED_OPST_DEAD:
 			/* Unrecoverable self-test error */
-			return TEE_ERROR_BAD_STATE;
+			return false;
+		case SEED_OPST_BIST:
+		case SEED_OPST_WAIT:
 		default:
-			break; /* can't happen */
+			cpu_relax();
+			continue;
+		}
+	} while (!timeout_elapsed(timeout));
+
+	EMSG("Failed to produce a sufficient amount of entropy");
+
+	return false;
+}
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+	uint8_t *ptr = buf;
+	uint16_t seed = 0;
+
+	while (len > 0) {
+		if (!seed_get_random_u16(&seed))
+			return TEE_ERROR_ACCESS_DENIED;
+		*ptr++ = seed & 0xff;
```

Comment:
> Actually, don't you skip the first 8 bytes of the buffer here? As ++ happens before the assignation.

After checking generated assembly code, I can confirm that the assignment (store instruction) is executed on correct address instead of (address + 1).
Do you concern that different compilers might generate different program order ?

Reply:
First, I swapped "bytes" for "bits", then I missed that the pointer is incremented after the assignation so please, discard the first part of my comment :)

Reply:
Got it! no problem :)

Reply:
> can you proceed by adding commits [Review] <commit title> that can be squashed after the tag is given? This ensure that no other modifications are made to the original commit and simplifies the review :)

OK, will do.